src/Symfony/Bundle/TwigBundle/TwigEngine.php

```diff
@@ -74,10 +74,15 @@ public function render($name, array $parameters = array())
             if ($name instanceof TemplateReference) {
                 try {
                     // try to get the real name of the template where the error occurred
-                    $name = $e->getTemplateName();
-                    $path = (string) $this->locator->locate($this->parser->parse($name));
+                    if (method_exists($e, 'getSourceContext')) {
+                        $templateName = $e->getSourceContext()->getName();
+                        $path = $e->getSourceContext()->getPath();
+                    } else {
+                        $templateName = $e->getTemplateName();
+                        $path = (string) $this->locator->locate($this->parser->parse($templateName));
+                    }
                     if (method_exists($e, 'setSourceContext')) {
-                        $e->setSourceContext(new \Twig_Source('', $name, $path));
+                        $e->setSourceContext(new \Twig_Source('', $templateName, $path));
```

Comment:
After this @attila901 will end up something like this:

```php
if (method_exists($e, 'setSourceContext')) {
    $e->setSourceContext($e->getSourceContext());
} else {
    $templateName = $e->getTemplateName();
    $path = (string) $this->locator->locate($this->parser->parse($templateName));
    $e->setTemplateName($path);
}
```

The ```$e->setSourceContext($e->getSourceContext());``` part can be skiped and we are not using the ```$name``` variable at all. Maybe after introducing ```setSourceContext```/```getSourceContext``` it is always filled or we should get necessary info  based on ```$name```?